internal/consensus/state.go

```diff
@@ -1018,22 +1028,28 @@ func (cs *State) handleMsg(ctx context.Context, mi msgInfo, fsyncUponCompletion
 
 	switch msg := msg.(type) {
 	case *ProposalMessage:
-		_, span := cs.tracer.Start(cs.getTracingCtx(ctx), "cs.state.handleProposalMsg")
+		spanCtx, span := cs.tracer.Start(cs.getTracingCtx(ctx), "cs.state.handleProposalMsg")
 		span.SetAttributes(attribute.Int("round", int(msg.Proposal.Round)))
 		defer span.End()
 
 		// will not cause transition.
 		// once proposal is set, we can receive block parts
 		err = cs.setProposal(msg.Proposal, mi.ReceiveTime)
+		// See if we can try creating the proposal block if keys exist
+		if cs.config.GossipTransactionKeyOnly && !cs.isProposer(cs.privValidatorPubKey.Address()) && cs.ProposalBlock == nil {
+			created := cs.tryCreateProposalBlock(spanCtx, msg.Proposal.Height, msg.Proposal.Round, msg.Proposal.Header, msg.Proposal.LastCommit, msg.Proposal.Evidence, msg.Proposal.ProposerAddress)
+			if created {
+				cs.fsyncAndCompleteProposal(ctx, fsyncUponCompletion, msg.Proposal.Height, span)
+			}
+		}
 
 	case *BlockPartMessage:
```

Comment:
one implication of still sending all block parts is that since each incoming message has to be processed sequentially (due to the mutex acquired at the beginning), the block parts message may block vote messages for prevotes (since the state will enter prevote if it's not missing any tx). It's not too much of a problem normally since we have optimistic processing, but if OP is ineligible (e.g. takes more than 1 round), then it may effectively degrade to the original logic (i.e. block part based proposals) even if the node is not missing any tx in the proposal.

Not a blocker by any means though

Reply:
hmmm good point! i think we can change it so that if we can proactively create the proposal already, in `case *BlockPartMessage` we can check if cs.Proposal is already set for this block part, then short circuit return. wdyt? at least the mutex won't be held for so long

Reply:
yeah i think that would work